File: src/cli/src/session.ts
import { AcpClient, ExecutionResult, PromptResult, SessionUpdateParams } from "@/acp/client.js";
import { cwd } from "process";

export class MustangSession {
  constructor(
    private client: AcpClient,
    public readonly sessionId: string,
  ) {}

  static async create(
    client: AcpClient,
    workingDir?: string,
  ): Promise<MustangSession> {
    const result = await client.request<{ sessionId: string }>("session/new", {
      cwd: workingDir ?? cwd(),
      mcpServers: [],
    });
    return new MustangSession(client, result.sessionId);
  }

  static async load(
    client: AcpClient,
    id: string,
    workingDir?: string,
  ): Promise<MustangSession> {
    await client.request("session/load", {
      sessionId: id,
      cwd: workingDir ?? cwd(),
      mcpServers: [],
    });
    return new MustangSession(client, id);
  }

  async prompt(
    text: string,
    onUpdate: (update: SessionUpdateParams) => void,
  ): Promise<PromptResult> {
    const unsub = this.client.onUpdate(onUpdate);
    try {
      return await this.client.promptRequest(this.sessionId, text);
    } finally {
      unsub();
    }
  }

  async executeShell(
    command: string,
    excludeFromContext: boolean,
    onUpdate: (update: SessionUpdateParams) => void,
  ): Promise<ExecutionResult> {
    const unsub = this.client.onUpdate(onUpdate);
    try {
      return await this.client.executeShellRequest(this.sessionId, command, excludeFromContext);
    } finally {
      unsub();
    }
  }

  async executePython(
    code: string,
    excludeFromContext: boolean,
    onUpdate: (update: SessionUpdateParams) => void,
  ): Promise<ExecutionResult> {
    const unsub = this.client.onUpdate(onUpdate);
    try {
      return await this.client.executePythonRequest(this.sessionId, code, excludeFromContext);
    } finally {
      unsub();
    }
  }

  cancel(): void {
    this.client.notify("session/cancel", { sessionId: this.sessionId });
  }

  cancelExecution(kind: "shell" | "python" | "any" = "any"): void {
    this.client.notify("session/cancel_execution", { sessionId: this.sessionId, kind });
  }

  async setMode(mode: "default" | "plan"): Promise<void> {
    await this.client.request("session/set_mode", {
      sessionId: this.sessionId,
      modeId: mode,
    });
  }
}

File: src/kernel/kernel/protocol/acp/schemas/session.py
"""ACP session method wire-format schemas."""

from __future__ import annotations

from typing import Any
from typing import Literal

from kernel.protocol.acp.schemas.base import AcpModel
from kernel.protocol.acp.schemas.content import AcpContentBlock
from kernel.protocol.acp.schemas.enums import AcpStopReason


class AcpMcpServer(AcpModel):
    name: str
    command: str | None = None
    args: list[str] = []
    env: list[dict[str, str]] = []
    type: str | None = None
    url: str | None = None
    headers: list[dict[str, str]] = []


class AcpSessionInfo(AcpModel):
    session_id: str
    cwd: str
    updated_at: str
    title: str | None = None
    archived_at: str | None = None
    title_source: Literal["auto", "user"] | None = None
    meta: dict[str, Any] | None = None


# session/new


class NewSessionRequest(AcpModel):
    cwd: str
    mcp_servers: list[AcpMcpServer] = []
    meta: dict[str, Any] | None = None


class NewSessionResponse(AcpModel):
    session_id: str
    config_options: list[dict] | None = None
    modes: dict | None = None
    meta: dict[str, Any] | None = None


# session/load


class LoadSessionRequest(AcpModel):
    session_id: str
    cwd: str
    mcp_servers: list[AcpMcpServer] = []
    meta: dict[str, Any] | None = None


class LoadSessionResponse(AcpModel):
    config_options: list[dict] | None = None
    modes: dict | None = None
    meta: dict[str, Any] | None = None


# session/list


class ListSessionsRequest(AcpModel):
    cursor: str | None = None
    cwd: str | None = None
    include_archived: bool = False
    archived_only: bool = False
    meta: dict[str, Any] | None = None


class ListSessionsResponse(AcpModel):
    sessions: list[AcpSessionInfo]
    next_cursor: str | None = None
    meta: dict[str, Any] | None = None


# session/prompt


class PromptRequest(AcpModel):
    session_id: str
    prompt: list[AcpContentBlock]
    max_turns: int = 0
    meta: dict[str, Any] | None = None


class PromptResponse(AcpModel):
    stop_reason: AcpStopReason
    meta: dict[str, Any] | None = None


# session/execute_shell


class ExecuteShellRequest(AcpModel):
    session_id: str
    command: str
    exclude_from_context: bool = False
    shell: str = "auto"
    meta: dict[str, Any] | None = None


class ExecuteShellResponse(AcpModel):
    exit_code: int
    cancelled: bool = False
    meta: dict[str, Any] | None = None


# session/execute_python


class ExecutePythonRequest(AcpModel):
    session_id: str
    code: str
    exclude_from_context: bool = False
    meta: dict[str, Any] | None = None


class ExecutePythonResponse(AcpModel):
    exit_code: int
    cancelled: bool = False
    meta: dict[str, Any] | None = None


# session/cancel_execution


class CancelExecutionRequest(AcpModel):
    session_id: str
    kind: str = "any"
    meta: dict[str, Any] | None = None


class CancelExecutionResponse(AcpModel):
    meta: dict[str, Any] | None = None


# session/cancel (notification)


class CancelNotification(AcpModel):
    session_id: str
    meta: dict[str, Any] | None = None


class CancelRequestNotification(AcpModel):
    request_id: str | int
    meta: dict[str, Any] | None = None


# session/set_mode


class SetSessionModeRequest(AcpModel):
    session_id: str
    mode_id: str
    meta: dict[str, Any] | None = None


class SetSessionModeResponse(AcpModel):
    meta: dict[str, Any] | None = None


# session/set_config_option


class SetSessionConfigOptionRequest(AcpModel):
    session_id: str
    config_id: str
    value: str
    meta: dict[str, Any] | None = None


class SetSessionConfigOptionResponse(AcpModel):
    config_options: list[dict]
    meta: dict[str, Any] | None = None


# session/rename


class RenameSessionRequest(AcpModel):
    session_id: str
    title: str
    meta: dict[str, Any] | None = None


class RenameSessionResponse(AcpModel):
    session: AcpSessionInfo
    meta: dict[str, Any] | None = None


# session/archive


class ArchiveSessionRequest(AcpModel):
    session_id: str
    archived: bool = True
    meta: dict[str, Any] | None = None


class ArchiveSessionResponse(AcpModel):
    session: AcpSessionInfo
    meta: dict[str, Any] | None = None


# session/delete


class DeleteSessionRequest(AcpModel):
    session_id: str
    force: bool = False
    meta: dict[str, Any] | None = None


class DeleteSessionResponse(AcpModel):
    deleted: bool
    meta: dict[str, Any] | None = None
